coderd/jwtutils/jwe.go

```diff
@@ -0,0 +1,125 @@
+package jwtutils
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"time"
+
+	"github.com/go-jose/go-jose/v4"
+	"github.com/go-jose/go-jose/v4/jwt"
+	"golang.org/x/xerrors"
+)
+
+const (
+	encryptKeyAlgo     = jose.A256GCMKW
+	encryptContentAlgo = jose.A256GCM
+)
+
+type EncryptKeyProvider interface {
+	EncryptingKey(ctx context.Context) (id string, key interface{}, err error)
+}
+
+type DecryptKeyProvider interface {
+	DecryptingKey(ctx context.Context, id string) (key interface{}, err error)
+}
+
+// Encrypt encrypts a token and returns it as a string.
+func Encrypt(ctx context.Context, e EncryptKeyProvider, claims Claims) (string, error) {
+	id, key, err := e.EncryptingKey(ctx)
+	if err != nil {
+		return "", xerrors.Errorf("get signing key: %w", err)
+	}
+
+	encrypter, err := jose.NewEncrypter(
+		encryptContentAlgo,
+		jose.Recipient{
+			Algorithm: encryptKeyAlgo,
+			Key:       key,
+		},
+		&jose.EncrypterOptions{
+			Compression: jose.DEFLATE,
+			ExtraHeaders: map[jose.HeaderKey]interface{}{
+				keyIDHeaderKey: id,
+			},
+		},
+	)
+	if err != nil {
+		return "", xerrors.Errorf("initialize encrypter: %w", err)
+	}
+
+	payload, err := json.Marshal(claims)
+	if err != nil {
+		return "", xerrors.Errorf("marshal payload: %w", err)
+	}
+
+	encrypted, err := encrypter.Encrypt(payload)
+	if err != nil {
+		return "", xerrors.Errorf("encrypt: %w", err)
+	}
+
+	serialized := []byte(encrypted.FullSerialize())
+	return base64.RawURLEncoding.EncodeToString(serialized), nil
+}
+
+// DecryptOptions are options for decrypting a JWE.
+type DecryptOptions struct {
+	RegisteredClaims jwt.Expected
+
+	// The following should only be used for JWEs.
+	KeyAlgorithm               jose.KeyAlgorithm
+	ContentEncryptionAlgorithm jose.ContentEncryption
+}
+
+// Decrypt decrypts the token using the provided key. It unmarshals into the provided claims.
+func Decrypt(ctx context.Context, d DecryptKeyProvider, token string, claims Claims, opts ...func(*DecryptOptions)) error {
+	options := DecryptOptions{
+		RegisteredClaims: jwt.Expected{
+			Time: time.Now(),
+		},
+		KeyAlgorithm:               encryptKeyAlgo,
+		ContentEncryptionAlgorithm: encryptContentAlgo,
+	}
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	encrypted, err := base64.RawURLEncoding.DecodeString(token)
+	if err != nil {
+		return xerrors.Errorf("decode: %w", err)
+	}
+
+	object, err := jose.ParseEncrypted(string(encrypted),
```

Comment:
The inverse of above. If you do `base64.RawURLEncoding`, then you have to manually base64 it back into `json`. But we probably want the compact base64 format, then we do not need to use the `base64` package at all.